File: search/d972_packed_gf3_echelon_backend_v1.c
/* D972 packed GF(3) echelon backend candidate.  Portable C11, schema v1. */
#include <errno.h>
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAGIC "D972P3GF"
#define MAGIC_BYTES 8u
#define HEADER_BYTES 48u
#define SCHEMA_NAME "packed-gf3-echelon-v1"
#define MAX_WIDTH 10000000ULL
#define MAX_ROWS 100000ULL
#define MAX_LEDGER 10000000ULL
#define MAX_STATE_BYTES (512ULL * 1024ULL * 1024ULL)

typedef struct { uint64_t pivot, coefficient; } Reduction;
typedef struct {
    uint64_t id;
    uint8_t *row;
    uint64_t lead;
} Basis;
typedef struct {
    uint64_t id;
    Reduction *red;
    uint64_t nred, capred;
    int accepted;
    uint64_t pivot, lead, leading_coefficient, scale;
} Offered;

static int add_ok(uint64_t a, uint64_t b) { return b <= UINT64_MAX - a; }
static int mul_ok(uint64_t a, uint64_t b) { return a == 0 || b <= UINT64_MAX / a; }
static int read_exact(FILE *f, void *p, size_t n) { return n == 0 || fread(p, 1, n, f) == n; }
static uint32_t le32(const uint8_t *p) { return (uint32_t)p[0] | ((uint32_t)p[1] << 8) | ((uint32_t)p[2] << 16) | ((uint32_t)p[3] << 24); }
static uint64_t le64(const uint8_t *p) {
    uint64_t x = 0; unsigned i;
    for (i = 0; i < 8; ++i) x |= ((uint64_t)p[i]) << (8u * i);
    return x;
}
static void put_u64(uint8_t *p, uint64_t x) {
    unsigned i; for (i = 0; i < 8; ++i) p[i] = (uint8_t)(x >> (8u * i));
}
static int trit(const uint8_t *row, uint64_t coord) {
    return (int)((row[coord / 4u] / (uint8_t)(coord % 4u == 0 ? 1 : coord % 4u == 1 ? 3 : coord % 4u == 2 ? 9 : 27)) % 3u);
}
static uint8_t pack4(const int *v) { return (uint8_t)(v[0] + 3*v[1] + 9*v[2] + 27*v[3]); }
static uint8_t axpy_byte(uint8_t a, uint8_t c, uint8_t b) {
    int x[4], y[4], z[4], i;
    for (i = 0; i < 4; ++i) { x[i] = (a / (i == 0 ? 1 : i == 1 ? 3 : i == 2 ? 9 : 27)) % 3; y[i] = (b / (i == 0 ? 1 : i == 1 ? 3 : i == 2 ? 9 : 27)) % 3; z[i] = (x[i] - (int)c*y[i]) % 3; if (z[i] < 0) z[i] += 3; }
    return pack4(z);
}
static int first_trit(uint8_t b) { int i; for (i = 0; i < 4; ++i) if (trit(&b, (uint64_t)i)) return i; return -1; }
static int row_invariant(const uint8_t *row, uint64_t packed, uint64_t lead) {
    uint64_t c;
    if (lead >= packed * 4u || trit(row, lead) != 1) return 0;
    for (c = 0; c < lead; ++c) if (trit(row, c) != 0) return 0;
    return 1;
}
static int append_red(Offered *o, uint64_t pivot, uint64_t coefficient, uint64_t *ledger_total) {
    Reduction *p; uint64_t ncap;
    if (o->nred >= MAX_LEDGER || *ledger_total >= MAX_LEDGER) return 0;
    if (o->nred == o->capred) {
        ncap = o->capred ? o->capred * 2u : 8u;
        if (ncap < o->capred || ncap > MAX_LEDGER || !mul_ok(ncap, sizeof(*p))) return 0;
        p = (Reduction *)realloc(o->red, (size_t)(ncap * sizeof(*p))); if (!p) return 0;
        o->red = p; o->capred = ncap;
    }
    o->red[o->nred].pivot = pivot; o->red[o->nred].coefficient = coefficient; ++o->nred; ++*ledger_total;
    return 1;
}
static int reduce(const uint8_t *input, uint8_t *work, uint64_t packed, int64_t *lead_to_pivot, Basis *basis, Offered *o, uint64_t *ledger_total) {
    uint64_t cursor, j, lead, pivot; int off, coeff;
    memcpy(work, input, (size_t)packed);
    for (cursor = 0; cursor < packed;) {
        if (work[cursor] == 0) { ++cursor; continue; }
        off = first_trit(work[cursor]); if (off < 0) return 0;
        lead = cursor * 4u + (uint64_t)off; pivot = (uint64_t)(lead_to_pivot[lead]);
        if (lead_to_pivot[lead] < 0) break;
        coeff = trit(work, lead); if (coeff != 1 && coeff != 2) return 0;
        if (!append_red(o, pivot, (uint64_t)coeff, ledger_total)) return 0;
        /* The invariant is checked on every pivot before suffix update. */
        if (!row_invariant(basis[pivot].row, packed, basis[pivot].lead)) return 0;
        for (j = cursor; j < packed; ++j) work[j] = axpy_byte(work[j], (uint8_t)coeff, basis[pivot].row[j]);
    }
    return 1;
}
static int normalize(uint8_t *row, uint64_t packed, uint64_t *lead, uint64_t *lc, uint64_t *scale) {
    uint64_t b; int off, i, v[4];
    for (b = 0; b < packed && row[b] == 0; ++b) {}
    if (b == packed) return 0;
    off = first_trit(row[b]); if (off < 0) return 0;
    *lead = b * 4u + (uint64_t)off; *lc = (uint64_t)trit(row, *lead); *scale = *lc == 1 ? 1u : 2u;
    if (*scale == 2u) for (b = 0; b < packed; ++b) { for (i = 0; i < 4; ++i) v[i] = (row[b] / (i == 0 ? 1 : i == 1 ? 3 : i == 2 ? 9 : 27)) % 3; for (i = 0; i < 4; ++i) v[i] = (2*v[i]) % 3; row[b] = pack4(v); }
    return 1;
}
static void json_bytes(FILE *out, const uint8_t *row, uint64_t n) {
    uint64_t i; fputc('[', out); for (i = 0; i < n; ++i) { if (i) fputc(',', out); fprintf(out, "%u", (unsigned)row[i]); } fputc(']', out);
}
static void json_red(FILE *out, const Reduction *r, uint64_t n) {
    uint64_t i; fputc('[', out); for (i = 0; i < n; ++i) { if (i) fputc(',', out); fprintf(out, "[%" PRIu64 ",%" PRIu64 "]", r[i].pivot, r[i].coefficient); } fputc(']', out);
}
static void free_offered(Offered *o, uint64_t n) { uint64_t i; if (!o) return; for (i = 0; i < n; ++i) free(o[i].red); free(o); }
static void usage(void) { fprintf(stderr, "usage: backend --version 1 --schema " SCHEMA_NAME " --input FILE --output FILE\n"); }

int main(int argc, char **argv) {
    const char *input_path = NULL, *output_path = NULL, *version = NULL, *schema = NULL;
    FILE *in = NULL, *out = NULL; uint8_t h[HEADER_BYTES], *work = NULL, *target = NULL;
    Basis *basis = NULL; Offered *off = NULL; int64_t *lead_to_pivot = NULL; uint64_t width, packed, nrows, i, j, id;
    uint32_t hv, hs; uint64_t accepted = 0, lead, lc, scale, ledger_total = 0; int rc = 2; Offered target_o;
    memset(&target_o, 0, sizeof(target_o));
    for (i = 1; i < (uint64_t)argc; ++i) {
        if (strcmp(argv[i], "--input") == 0 && i+1 < (uint64_t)argc) input_path = argv[++i];
        else if (strcmp(argv[i], "--output") == 0 && i+1 < (uint64_t)argc) output_path = argv[++i];
        else if (strcmp(argv[i], "--version") == 0 && i+1 < (uint64_t)argc) version = argv[++i];
        else if (strcmp(argv[i], "--schema") == 0 && i+1 < (uint64_t)argc) schema = argv[++i];
        else { usage(); goto done; }
    }
    if (!input_path || !output_path || !version || !schema || strcmp(version, "1") || strcmp(schema, SCHEMA_NAME)) { usage(); goto done; }
    in = fopen(input_path, "rb"); if (!in) goto done;
    if (!read_exact(in, h, sizeof(h)) || memcmp(h, MAGIC, MAGIC_BYTES) != 0) goto done;
    hv = le32(h+8); hs = le32(h+12); width = le64(h+16); nrows = le64(h+24); packed = le64(h+32);
    if (hv != 1u || hs != 1u || le64(h+40) != 0 || width == 0 || width > MAX_WIDTH || width % 4u || packed != width/4u || nrows > MAX_ROWS) goto done;
    if (!mul_ok(packed, sizeof(uint8_t)) || packed > (uint64_t)SIZE_MAX || !mul_ok(width, sizeof(*lead_to_pivot)) || width * sizeof(*lead_to_pivot) > (uint64_t)SIZE_MAX || !mul_ok(nrows, sizeof(*basis)) || nrows * sizeof(*basis) > (uint64_t)SIZE_MAX || !mul_ok(nrows, sizeof(*off)) || nrows * sizeof(*off) > (uint64_t)SIZE_MAX) goto done;
    /* Check the complete declared payload size before any allocation/read. */
    if (!mul_ok(nrows, 8u + packed) || !add_ok(HEADER_BYTES, nrows * (8u + packed)) || !add_ok(HEADER_BYTES + nrows * (8u + packed), packed) || !mul_ok(nrows, packed) || nrows * packed > MAX_STATE_BYTES) goto done;
    lead_to_pivot = (int64_t *)malloc((size_t)(width * sizeof(*lead_to_pivot))); if (!lead_to_pivot) goto done;
    for (i = 0; i < width; ++i) lead_to_pivot[i] = -1;
    work = (uint8_t *)malloc((size_t)packed); target = (uint8_t *)malloc((size_t)packed);
    basis = (Basis *)calloc((size_t)(nrows ? nrows : 1u), sizeof(*basis)); off = (Offered *)calloc((size_t)(nrows ? nrows : 1u), sizeof(*off));
    if (!work || !target || !basis || !off) goto done;
    for (i = 0; i < nrows; ++i) {
        uint8_t idbuf[8]; off[i].red = NULL;
        if (!read_exact(in, idbuf, 8) || !read_exact(in, work, (size_t)packed)) goto done;
        id = le64(idbuf); off[i].id = id; for (j = 0; j < packed; ++j) if (work[j] > 80u) goto done;
        if (!reduce(work, work, packed, lead_to_pivot, basis, &off[i], &ledger_total)) goto done;
        if (!normalize(work, packed, &lead, &lc, &scale)) { off[i].accepted = 0; continue; }
        if (lead_to_pivot[lead] >= 0 || !row_invariant(work, packed, lead)) goto done;
        off[i].accepted = 1; off[i].pivot = accepted; off[i].lead = lead; off[i].leading_coefficient = lc; off[i].scale = scale;
        basis[accepted].id = id; basis[accepted].lead = lead; basis[accepted].row = (uint8_t *)malloc((size_t)packed); if (!basis[accepted].row) goto done; memcpy(basis[accepted].row, work, (size_t)packed);
        lead_to_pivot[lead] = (int64_t)accepted; ++accepted;
    }
    if (!read_exact(in, target, (size_t)packed)) goto done; for (j = 0; j < packed; ++j) if (target[j] > 80u) goto done;
    if (fgetc(in) != EOF || ferror(in)) goto done;
    if (!reduce(target, target, packed, lead_to_pivot, basis, &target_o, &ledger_total)) goto done;
    out = fopen(output_path, "wb"); if (!out) goto done;
    fprintf(out, "{\"version\":1,\"schema\":\"" SCHEMA_NAME "\",\"width\":%" PRIu64 ",\"packed_bytes\":%" PRIu64 ",\"accepted_basis\":[", width, packed);
    for (i = 0; i < accepted; ++i) { if (i) fputc(',', out); fprintf(out, "{\"pivot\":%" PRIu64 ",\"row_id\":%" PRIu64 ",\"lead\":%" PRIu64 ",\"bytes\":", i, basis[i].id, basis[i].lead); json_bytes(out, basis[i].row, packed); fputc('}', out); }
    fprintf(out, "],\"offered\":[");
    for (i = 0; i < nrows; ++i) { if (i) fputc(',', out); fprintf(out, "{\"row_id\":%" PRIu64 ",\"reductions\":", off[i].id); json_red(out, off[i].red, off[i].nred); fprintf(out, ",\"accepted\":%s", off[i].accepted ? "true" : "false"); if (off[i].accepted) fprintf(out, ",\"pivot\":%" PRIu64 ",\"lead\":%" PRIu64 ",\"leading_coefficient\":%" PRIu64 ",\"scale\":%" PRIu64, off[i].pivot, off[i].lead, off[i].leading_coefficient, off[i].scale); fputc('}', out); }
    fprintf(out, "],\"target\":{\"reductions\":"); json_red(out, target_o.red, target_o.nred); fprintf(out, ",\"coefficients\":"); json_red(out, target_o.red, target_o.nred); fprintf(out, ",\"remainder\":"); json_bytes(out, target, packed); fprintf(out, "}}\n");
    if (fclose(out) != 0) { out = NULL; goto done; } out = NULL; rc = 0;
done:
    if (out) fclose(out); if (in) fclose(in); for (i = 0; i < accepted; ++i) free(basis[i].row); free(basis); free_offered(off, nrows); free(target_o.red); free(work); free(target); free(lead_to_pivot); return rc;
}

File: search/d972_packed_gf3_echelon_backend_v1.py
"""Strict ABI wrapper and small reference for the packed GF(3) worker.

The wrapper never silently falls back to Python when a compiled executable is
absent.  Python reference functions are intentionally useful only for tests.
"""
from __future__ import annotations

import json
import os
import struct
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterable, Sequence

MAGIC = b"D972P3GF"
VERSION = 1
SCHEMA = "packed-gf3-echelon-v1"
HEADER = struct.Struct("<8sIIQQQQ")
MAX_WIDTH = 10_000_000
MAX_ROWS = 100_000
MAX_LEDGER = 10_000_000


class BackendError(ValueError):
    pass


class BackendUnavailable(RuntimeError):
    pass


def _check_shape(width: int, rows: Sequence[Sequence[int]], target: Sequence[int], ids: Sequence[int]) -> int:
    if not isinstance(width, int) or width <= 0 or width % 4 or width > MAX_WIDTH:
        raise BackendError("width")
    if len(rows) != len(ids) or len(rows) > MAX_ROWS or len(target) != width:
        raise BackendError("row_count")
    for row in (*rows, target):
        if len(row) != width or any(type(x) is not int or x not in (0, 1, 2) for x in row):
            raise BackendError("dense_row")
    if any(type(x) is not int or x < 0 or x >= 2**64 for x in ids):
        raise BackendError("row_id")
    return width // 4


def pack_row(row: Sequence[int]) -> bytes:
    if len(row) % 4:
        raise BackendError("pack_width")
    out = bytearray(len(row) // 4)
    for b in range(len(out)):
        x = 0
        for j in range(4):
            v = row[4 * b + j]
            if type(v) is not int or v not in (0, 1, 2):
                raise BackendError("trit")
            x += v * (3**j)
        out[b] = x
    return bytes(out)


def unpack_row(packed: bytes, width: int | None = None) -> list[int]:
    if width is None:
        width = len(packed) * 4
    if width <= 0 or width % 4 or len(packed) != width // 4 or any(x > 80 for x in packed):
        raise BackendError("packed_row")
    out: list[int] = []
    for x in packed:
        for j in range(4):
            out.append((x // (3**j)) % 3)
    return out


def write_input(path: str | os.PathLike[str], width: int, rows: Sequence[Sequence[int]], target: Sequence[int], row_ids: Sequence[int] | None = None) -> None:
    ids = list(range(len(rows))) if row_ids is None else list(row_ids)
    packed = _check_shape(width, rows, target, ids)
    with open(path, "wb") as f:
        f.write(HEADER.pack(MAGIC, VERSION, 1, width, len(rows), packed, 0))
        for row_id, row in zip(ids, rows):
            f.write(struct.pack("<Q", row_id))
            f.write(pack_row(row))
        f.write(pack_row(target))


def run_compiled(input_path: str | os.PathLike[str], receipt_path: str | os.PathLike[str], executable: str | os.PathLike[str] | None = None) -> dict[str, Any]:
    """Run only an explicitly selected compiled worker; missing worker fails closed."""
    exe = str(executable or os.environ.get("D972_PACKED_GF3_BACKEND", ""))
    if not exe:
        raise BackendUnavailable("compiled_backend_not_configured")
    if not Path(exe).is_file():
        raise BackendUnavailable("compiled_backend_missing")
    cmd = [exe, "--version", "1", "--schema", SCHEMA, "--input", str(input_path), "--output", str(receipt_path)]
    try:
        cp = subprocess.run(cmd, check=False, capture_output=True, text=True, timeout=30)
    except (OSError, subprocess.TimeoutExpired) as exc:
        raise BackendUnavailable("compiled_backend_failed") from exc
    if cp.returncode != 0:
        raise BackendError("compiled_backend_rejected")
    return parse_receipt(receipt_path)


def _first(row: bytearray) -> int | None:
    for b, x in enumerate(row):
        if x:
            for j in range(4):
                if (x // (3**j)) % 3:
                    return 4 * b + j
    return None


def _axpy(a: int, c: int, b: int) -> int:
    z = 0
    for j in range(4):
        v = ((a // (3**j)) % 3 - c * ((b // (3**j)) % 3)) % 3
        z += v * (3**j)
    return z


def _scale2(a: int) -> int:
    z = 0
    for j in range(4):
        z += (2 * ((a // (3**j)) % 3) % 3) * (3**j)
    return z


@dataclass
class ReferenceEchelon:
    width: int
    rows: list[bytes]
    leads: list[int]
    row_ids: list[int]

    def __init__(self, width: int):
        if width <= 0 or width % 4:
            raise BackendError("width")
        self.width, self.rows, self.leads, self.row_ids = width, [], [], []

    def reduce(self, packed: bytes) -> tuple[bytes, list[list[int]]]:
        if len(packed) != self.width // 4 or any(x > 80 for x in packed):
            raise BackendError("packed_row")
        work = bytearray(packed)
        reductions: list[list[int]] = []
        while True:
            lead = _first(work)
            if lead is None:
                break
            try:
                pivot = self.leads.index(lead)
            except ValueError:
                break
            coeff = (work[lead // 4] // (3 ** (lead % 4))) % 3
            reductions.append([pivot, coeff])
            pivot_row = self.rows[pivot]
            if any(unpack_row(pivot_row, self.width)[: self.leads[pivot]]):
                raise BackendError("pivot_invariant")
            start = lead // 4
            for j in range(start, len(work)):
                work[j] = _axpy(work[j], coeff, pivot_row[j])
        return bytes(work), reductions

    def insert(self, row: Sequence[int], row_id: int = 0) -> dict[str, Any]:
        packed = pack_row(row)
        rem, reductions = self.reduce(packed)
        lead = _first(bytearray(rem))
        result: dict[str, Any] = {"row_id": row_id, "reductions": reductions, "accepted": lead is not None}
        if lead is None:
            return result
        lc = (rem[lead // 4] // (3 ** (lead % 4))) % 3
        scale = 1 if lc == 1 else 2
        norm = bytes(_scale2(x) if scale == 2 else x for x in rem)
        if lead in self.leads:
            raise BackendError("duplicate_lead")
        if any(unpack_row(norm, self.width)[:lead]) or unpack_row(norm, self.width)[lead] != 1:
            raise BackendError("pivot_invariant")
        pivot = len(self.rows)
        self.rows.append(norm); self.leads.append(lead); self.row_ids.append(row_id)
        result.update(pivot=pivot, lead=lead, leading_coefficient=lc, scale=scale)
        return result

    def receipt(self, rows: Sequence[Sequence[int]], ids: Sequence[int], target: Sequence[int]) -> dict[str, Any]:
        offered = [self.insert(row, int(row_id)) for row, row_id in zip(rows, ids)]
        rem, reductions = self.reduce(pack_row(target))
        return {"version": 1, "schema": SCHEMA, "width": self.width, "packed_bytes": self.width // 4,
                "accepted_basis": [{"pivot": i, "row_id": self.row_ids[i], "lead": self.leads[i], "bytes": list(self.rows[i])} for i in range(len(self.rows))],
                "offered": offered, "target": {"reductions": reductions, "coefficients": reductions, "remainder": list(rem)}}


def reference_receipt(width: int, rows: Sequence[Sequence[int]], target: Sequence[int], row_ids: Sequence[int] | None = None) -> dict[str, Any]:
    ids = list(range(len(rows))) if row_ids is None else list(row_ids)
    _check_shape(width, rows, target, ids)
    return ReferenceEchelon(width).receipt(rows, ids, target)


def _pairs(value: Any) -> list[list[int]]:
    if type(value) is not list or any(type(x) is not list or len(x) != 2 or any(type(y) is not int or y < 0 for y in x) for x in value):
        raise BackendError("reductions")
    return value


def parse_receipt(path: str | os.PathLike[str]) -> dict[str, Any]:
    try:
        with open(path, "rb") as f:
            raw = f.read()
        obj = json.loads(raw.decode("utf-8"))
    except (OSError, UnicodeError, json.JSONDecodeError) as exc:
        raise BackendError("receipt_json") from exc
    if type(obj) is not dict or obj.get("version") != 1 or obj.get("schema") != SCHEMA:
        raise BackendError("receipt_schema")
    width, packed = obj.get("width"), obj.get("packed_bytes")
    if type(width) is not int or width <= 0 or width % 4 or type(packed) is not int or packed != width // 4 or width > MAX_WIDTH:
        raise BackendError("receipt_dimensions")
    basis = obj.get("accepted_basis"); offered = obj.get("offered"); target = obj.get("target")
    if type(basis) is not list or type(offered) is not list or len(offered) > MAX_ROWS or type(target) is not dict:
        raise BackendError("receipt_sections")
    leads: set[int] = set()
    for i, item in enumerate(basis):
        if type(item) is not dict or item.get("pivot") != i or type(item.get("row_id")) is not int or not (0 <= item["row_id"] < 2**64) or type(item.get("lead")) is not int:
            raise BackendError("basis_record")
        lead = item["lead"]; row = item.get("bytes")
        if lead < 0 or lead >= width or lead in leads or type(row) is not list or len(row) != packed or any(type(x) is not int or x < 0 or x > 80 for x in row):
            raise BackendError("basis_row")
        dense = unpack_row(bytes(row), width)
        if dense[lead] != 1 or any(dense[:lead]):
            raise BackendError("basis_invariant")
        leads.add(lead)
    for item in offered:
        if type(item) is not dict or type(item.get("row_id")) is not int or not (0 <= item["row_id"] < 2**64):
            raise BackendError("offered_record")
        pairs = _pairs(item.get("reductions"))
        if any(p[0] >= len(basis) or p[1] not in (1, 2) for p in pairs):
            raise BackendError("reduction_range")
        if type(item.get("accepted")) is not bool:
            raise BackendError("accepted_flag")
        if item["accepted"]:
            if type(item.get("pivot")) is not int or not (0 <= item["pivot"] < len(basis)) or type(item.get("lead")) is not int or item["lead"] not in leads or type(item.get("leading_coefficient")) is not int or item["leading_coefficient"] not in (1, 2) or type(item.get("scale")) is not int or item["scale"] not in (1, 2):
                raise BackendError("accepted_record")
    _pairs(target.get("reductions")); _pairs(target.get("coefficients"))
    if target["reductions"] != target["coefficients"]:
        raise BackendError("target_coefficients")
    rem = target.get("remainder")
    if type(rem) is not list or len(rem) != packed or any(type(x) is not int or x < 0 or x > 80 for x in rem):
        raise BackendError("target_remainder")
    return obj


__all__ = ["BackendError", "BackendUnavailable", "SCHEMA", "ReferenceEchelon", "pack_row", "unpack_row", "write_input", "run_compiled", "parse_receipt", "reference_receipt"]
